### PR TITLE
listFilesRecursive: use local function for recursion

### PR DESCRIPTION
**Pull request by milahu**
###### Description of changes

make the function easier to refactor / reuse

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by milahu** [comment]
similar function

useful to avoid IFD by processing files in Nix

```nix
{
  # findFilesRecursive: Path -> [ Path ]
  #
  # Given a directory, return a flattened list of all files
  # that meet the criteria in options.
  #
  # Examples:
  # findFilesRecursive /some/path { name = "readme.md"; maxDepth = 2; }
  # findFilesRecursive /some/path { name = "yarn.lock"; skipDirs = [ "node_modules" ]; }
  #
  findFilesRecursive = dir: options:
    let
      recurse = dir: depth:
        lib.flatten (lib.mapAttrsToList (name: type:
          if (options ? maxDepth && depth > options.maxDepth) then []
          else if type == "directory" then (
            if (options ? skipDirs && builtins.elem name options.skipDirs) then []
            else recurse (dir + "/${name}") (depth + 1)
          )
          else if (options ? name && name != options.name) then []
          else dir + "/${name}"
        ) (builtins.readDir dir));
    in
    recurse dir 0;
}
```

**Comment by milahu** [comment]
also useful ...

```nix
{

  # Get the relative file path of an absolute Nix store path.
  #
  # Example:
  #   storePath: /nix/store/x0klxz8xmxfz0x0pmi3d4c6kr11iwicc-source/build/npm/gyp/yarn.lock
  #   result: build/npm/gyp/yarn.lock

  getRelativeStorePath = storePath: (
    builtins.concatStringsSep "/" (lib.lists.drop 2 (
      lib.flatten (builtins.split "/" (removePrefixCheck builtins.storeDir storePath))
    ))
  );



  removePrefixCheck = prefix: str: (
    assert !(lib.hasPrefix prefix str) ->
      throw "removePrefixCheck: prefix '${prefix}' is missing in string '${str}'";
    lib.removePrefix prefix str
  );

}
```

File: lib/filesystem.nix [hunk 1]
```diff
@@ -47,11 +47,13 @@
...
   #
   # Given a directory, return a flattened list of all files within it recursively.
-  listFilesRecursive = dir: lib.flatten (lib.mapAttrsToList (name: type:
-    if type == "directory" then
-      lib.filesystem.listFilesRecursive (dir + "/${name}")
-    else
-      dir + "/${name}"
-  ) (builtins.readDir dir));
+  listFilesRecursive = dir: let
+    recurse = dir: lib.flatten (lib.mapAttrsToList (name: type:
```
Comment (roberth) [inline lib/filesystem.nix:51]: This `dir` shadows the outer `dir`, which isn't great.
The outer `dir` isn't even needed though, at which point you might as well remove the let binding, and we're back to square one.
To me, this implementation is more complicated and the steps required to reach this form are simple refactoring techniques. I don't think it's the purpose of this function's implementation to be something other than the simplest definition it can reasonably be.

**Comment by roberth** [comment]
> also useful ...
> 
> ```nix
> {
> 
>   # Get the relative file path of an absolute Nix store path.
>   #
>   # Example:
>   #   storePath: /nix/store/x0klxz8xmxfz0x0pmi3d4c6kr11iwicc-source/build/npm/gyp/yarn.lock
>   #   result: build/npm/gyp/yarn.lock

Not sure how this would be rendered. Many lib functions get rendered into the Nixpkgs manual.

> ```
>   getRelativeStorePath = storePath: (
>     builtins.concatStringsSep "/" (lib.lists.drop 2 (
>       lib.splitString "/" (removePrefixCheck builtins.storeDir storePath)
>     ))
>   );
> ```

Seems useful indeed, but the name sounds like it's going to return a store path. How about `splitStorePath`, which returns `{ storePath = "/nix/store/s0m3h4sh...-name"; subpath = "build/npm/gyp/yarn.lock"; }`.

> ```
>   removePrefixCheck = prefix: str: (
>     assert !(lib.hasPrefix prefix str) ->
> ```

Path prefix checking isn't string prefix checking. You'll want to use something like the following to avoid accidentally allowing substrings of path components: https://github.com/NixOS/nixpkgs/pull/112083/files#diff-c7ec15b333bb1ce2cb8065610ad1952b59fbb0fa244cf272c12b7e020c2fa146R75-R93

I'd like to hold back on adding lib functions of this sort because lazy-trees may give some constraints that - even if the code works - may cause a lazy tree to be overfetched.

**Comment by milahu** [comment]
> lazy-trees may give some constraints that - even if the code works - may cause a lazy tree to be overfetched.

ELI5 please : /

or was that a rephrase of the "compare paths" issue?

> #112083

> `normalizedPathString = pathLike: toString (/. + pathLike);`

nice!

> How about `splitStorePath`

i would add all 3 variants: dirname, basename, splitpath.
so users can optimize for their use case

**Comment by domenkozar** [comment]
user was banned https://github.com/NixOS/moderation/blob/main/moderation-log.md
